Lamina/src/GL/vertex_array.cpp

```diff
@@ -0,0 +1,82 @@
+#include "../Lamina/include/GL/vertex_array.hpp"
+
+void lm::VertexArray::DrawArray()
+{
+	try {
+		if (window == nullptr) std::runtime_error error("Window not assigned to vertex array\n");
+		if (camera == nullptr) std::runtime_error error("Camera not assigned to vertex array\n");
+	}
+	catch (std::runtime_error& exception)
+	{
+		std::cout << exception.what() << std::endl;
+		throw;
+		return;
+	}
+	catch (...) { throw; return; }
+
+	lm::ColorF glColor = lm::GetFloatColor(color);
+	lm::mat4 view = camera->GetCameraLookAtMatrix();
+
+	glBindVertexArray(VAO);
+	glBindBuffer(GL_ARRAY_BUFFER, VBO);
+	glActiveTexture(GL_TEXTURE0);
+	glUseProgram(program);
+
+	lm::mat4 perspectiveMatrix = lm::CreatePerspectiveProjectionMatrix(lm::constants::RadToDeg(70.f), (float)window->GetSize().x() / (float)window->GetSize().y(), 1, 10000);
+	glUniformMatrix4fv(glGetUniformLocation(program, "projectionMatrix"), 1, GL_TRUE, &perspectiveMatrix.data[0][0]);
+	glUniform4f(glGetUniformLocation(program, "color"), glColor.r, glColor.g, glColor.b, glColor.a);
+	glUniformMatrix4fv(glGetUniformLocation(program, "viewMatrix"), 1, GL_TRUE, &view.data[0][0]);
+	if (texture != nullptr)
+	{
+		glActiveTexture(GL_TEXTURE0);
+		glBindTexture(GL_TEXTURE_2D, texture->texture);
+		glUniform1i(glGetUniformLocation(program, "textured"), GL_TRUE);
+		glUniform1i(glGetUniformLocation(program, "inTexture"), 0);
+	}
+	else glUniform1i(glGetUniformLocation(program, "textured"), GL_FALSE);
+
+	glVertexAttribPointer(0, 4, GL_FLOAT, GL_FALSE, 9 * sizeof(GLfloat), (void*)0);
+	glVertexAttribPointer(1, 2, GL_FLOAT, GL_FALSE, 9 * sizeof(GLfloat), (void*)(4 * sizeof(GLfloat)));
+	glVertexAttribPointer(2, 3, GL_FLOAT, GL_FALSE, 9 * sizeof(GLfloat), (void*)(6 * sizeof(GLfloat)));
+
+	glEnableVertexAttribArray(0);
+	glEnableVertexAttribArray(1);
+	glEnableVertexAttribArray(2);
+
+	glDrawArrays(GL_TRIANGLES, 0, (GLsizei)mesh->meshData.size() / 9);
+
+	glDisableVertexAttribArray(0);
+	glDisableVertexAttribArray(1);
+	glDisableVertexAttribArray(2);
+}
+
+void lm::VertexArray::SetScale(float x, float y, float z)
+{
+	glUseProgram(program);
+	scale.CreateVector({ std::clamp(x, 0.f, (float)INT_MAX), std::clamp(y, 0.f, (float)INT_MAX), std::clamp(z, 0.f, (float)INT_MAX) });
```

Comment:
What is the motivation to clamp the value to `INT_MAX` which is the maximum for 32 bit integers? If you just want to prevent negative values, you can instead clamp to between `0.0f` and `FLT_MAX` which would be the actual upper limit in this case.